File: python/src/linodemcp/audit/event.py
"""Audit event vocabularies and the record they tag.

The record itself is the contract's own ``AuditEvent``, which declares
``local_record``, so the members and their order on disk come from the
declaration rather than from this language's dataclass. Mirrors
``go/internal/audit/event.go``.

The vocabularies below stay this module's own words: they are what the server
tags a call with, and the record carries their text.
"""

from __future__ import annotations

import secrets
from dataclasses import replace
from datetime import UTC, datetime
from enum import StrEnum
from typing import Any

from linodemcp.audit.redact import redact, redact_with_pii
from linodemcp.genlocal import AuditEvent

# Constant prefix on every event_id. Combined with a 26-char ULID
# body, the full id looks like ``evt_01HQXY3ZKQ8M7VRBNP4W5T2J9F``.
EVENT_ID_PREFIX = "evt_"

# Crockford base32 alphabet. I, L, O, U absent on purpose to avoid
# visual ambiguity with 1, 0, V.
_CROCKFORD_ALPHABET = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"


class Capability(StrEnum):
    """Capability tag, stringly-typed at the audit boundary.

    Mirrors the Go ``audit.Capability`` constants. Kept as strings
    rather than a richer enum so the JSON wire shape stays simple.
    """

    READ = "read"
    WRITE = "write"
    DESTROY = "destroy"
    ADMIN = "admin"
    META = "meta"


class Status(StrEnum):
    """Terminal status of a tool call.

    ``success`` is the happy path; ``error`` covers handler-level
    failures; ``refused`` covers profile blocks and validation
    failures (the handler never ran).
    """

    SUCCESS = "success"
    ERROR = "error"
    REFUSED = "refused"


class Mode(StrEnum):
    """Execution mode the call took.

    Default is ``normal``; dry-run + two-stage-writes specs
    introduce the remaining values.
    """

    NORMAL = "normal"
    DRY_RUN = "dry_run"
    PLAN = "plan"
    APPLY = "apply"
    BYPASS_DRY_RUN = "bypass" + "_dry_run"
    YOLO = "yolo"


# Event is one audit record per tool call, which is the contract's own
# AuditEvent under this module's name for it. One declaration owns the member
# set, their order and their types, so a log line reads the same whichever
# language wrote it.
Event = AuditEvent


def event_timestamp(instant: datetime) -> str:
    """One instant as an event's ts member spells it.

    No fractional part where the instant lands on a whole second, and
    microseconds where it does not, which is the spelling a record written by
    an earlier version carries and a reconstructed timestamp comes back in.
    """
    return instant.astimezone(UTC).isoformat().replace("+00:00", "Z")


def new_event(
    tool: str,
    capability: Capability,
    args: dict[str, Any],
    environment: str,
    profile: str,
    session_id: str,
    credential_generation: int,
    linodemcp_version: str,
    *,
    redact_pii: bool = False,
) -> Event:
    """Construct an Event with timestamp, ULID, and metadata populated.

    The remaining fields (status, latency, summary, error) populate
    later via :func:`finalize`. ``args`` is redacted in place: the
    returned event holds redacted args and the list of redacted keys.
    Callers that need the unredacted values keep their own copy.

    ``redact_pii`` controls the PII redaction tier (Phase 4c). The
    default False applies only the always-on credential list; the
    middleware passes True when the operator's config has
    ``audit.redact_pii: true`` (the default). Keyword-only to avoid
    accidental positional confusion with the other booleans the
    capture middleware might add later.

    The nanosecond count is built from the seconds and the microseconds rather
    than from a float, which loses the last digits of a modern timestamp.
    """
    now = datetime.now(UTC)
    redact_fn = redact_with_pii if redact_pii else redact
    redacted_args, redacted_keys = redact_fn(args)

    return AuditEvent(
        ts=event_timestamp(now),
        ts_unix_ns=int(now.timestamp()) * 1_000_000_000 + now.microsecond * 1000,
        event_id=new_event_id(now),
        tool=tool,
        tool_capability=capability.value,
        environment=environment,
        profile=profile,
        mode=Mode.NORMAL.value,
        plan_id=None,
        args=redacted_args if redacted_args is not None else {},
        args_redacted=redacted_keys,
        status=Status.SUCCESS.value,
        latency_ms=0,
        result_summary="",
        error=None,
        linodemcp_version=linodemcp_version,
        session_id=session_id,
        credential_generation=credential_generation,
    )


def finalize(
    event: Event,
    status: Status,
    latency_ms: int,
    err_msg: str,
    summary: str,
) -> Event:
    """Answer the event the call ended as.

    The capture middleware calls this once the handler returns. A new record
    rather than four writes into the old one, because the record is a value the
    contract owns and this language's is frozen. Empty ``err_msg`` clears the
    error to ``None`` so the record renders ``null``, not an empty string.
    """
    return replace(
        event,
        status=status.value,
        latency_ms=latency_ms,
        result_summary=summary,
        error=err_msg or None,
    )


def set_mode(event: Event, mode: Mode, plan_id: str) -> Event:
    """Answer the event under the execution mode the call took."""
    return replace(event, mode=mode.value, plan_id=plan_id or None)


def new_event_id(now: datetime) -> str:
    """Produce an EventID using ``now`` as the time component.

    Stateless: 80 bits of randomness make same-ms collisions
    negligible at the per-tool-call rate. Callers needing strict
    monotonic ordering should sort by ``Event.ts_unix_ns``, captured
    at the same instant.
    """
    timestamp_ms = int(now.timestamp() * 1000)
    entropy = secrets.token_bytes(10)
    return EVENT_ID_PREFIX + _encode_ulid(timestamp_ms, entropy)


def _encode_ulid(timestamp_ms: int, entropy: bytes) -> str:
    """Build the 26-char ULID body per the ULID spec."""
    time_len = 10
    random_len = 16
    bits_per_sym = 5

    out: list[str] = []

    # First 10 chars: 48-bit ms big-endian, base32-encoded.
    for i in range(time_len):
        shift = (time_len - 1 - i) * bits_per_sym
        out.append(_CROCKFORD_ALPHABET[(timestamp_ms >> shift) & 0x1F])

    # Remaining 16 chars: 80 bits of entropy, packed into 16 base32
    # symbols. The bit-twiddling matches the Go implementation
    # byte-for-byte so cross-language event IDs share the same shape.
    bits = [
        (entropy[0] & 0xF8) >> 3,
        ((entropy[0] & 0x07) << 2) | ((entropy[1] & 0xC0) >> 6),
        (entropy[1] & 0x3E) >> 1,
        ((entropy[1] & 0x01) << 4) | ((entropy[2] & 0xF0) >> 4),
        ((entropy[2] & 0x0F) << 1) | ((entropy[3] & 0x80) >> 7),
        (entropy[3] & 0x7C) >> 2,
        ((entropy[3] & 0x03) << 3) | ((entropy[4] & 0xE0) >> 5),
        entropy[4] & 0x1F,
        (entropy[5] & 0xF8) >> 3,
        ((entropy[5] & 0x07) << 2) | ((entropy[6] & 0xC0) >> 6),
        (entropy[6] & 0x3E) >> 1,
        ((entropy[6] & 0x01) << 4) | ((entropy[7] & 0xF0) >> 4),
        ((entropy[7] & 0x0F) << 1) | ((entropy[8] & 0x80) >> 7),
        (entropy[8] & 0x7C) >> 2,
        ((entropy[8] & 0x03) << 3) | ((entropy[9] & 0xE0) >> 5),
        entropy[9] & 0x1F,
    ]

    out.extend(_CROCKFORD_ALPHABET[value] for value in bits[:random_len])

    return "".join(out)

File: go/internal/audit/event.go
// Package audit defines the audit event schema and supporting
// helpers used by the audit-log spec. Phase 1a (this code) builds
// the event types and the redaction helper; Phase 1b wires capture
// into the tool dispatch middleware; later phases add JSONL and
// SQLite sinks plus query tools.
//
// The record itself is the contract's own AuditEvent, which declares
// local_record, so the members and their order on disk come from the
// declaration rather than from this language's struct layout. The
// vocabularies below stay this package's own words: they are what the
// server tags a call with, and the record carries their text.
package audit

import (
	"crypto/rand"
	"time"

	"github.com/chadit/LinodeMCP/go/internal/genlocal"
)

// EventIDPrefix is the constant prefix on every event_id. Combined
// with a 26-char ULID body, the full id looks like
// "evt_01HQXY3ZKQ8M7VRBNP4W5T2J9F".
const EventIDPrefix = "evt_"

// Capability mirrors profiles.Capability but stays a string for the
// audit wire format. Keeping it stringly-typed at the audit boundary
// avoids a cyclic dependency with the profiles package; callers
// translate from profiles.Capability via String() at event
// construction time.
type Capability string

// Capability constants. Lowercase to match the spec's wire format.
const (
	CapabilityRead    Capability = "read"
	CapabilityWrite   Capability = "write"
	CapabilityDestroy Capability = "destroy"
	CapabilityAdmin   Capability = "admin"
	CapabilityMeta    Capability = "meta"
)

// Status enumerates the terminal states of a tool call. `success`
// is the happy path; `error` covers handler-level failures; `refused`
// covers profile blocks and validation failures (i.e. the handler
// never ran).
type Status string

// Status constants.
const (
	StatusSuccess Status = "success"
	StatusError   Status = "error"
	StatusRefused Status = "refused"
)

// Mode enumerates the execution paths a call can take. Default is
// `normal`; the dry-run + two-stage-writes specs introduce the
// remaining values. The bypass value is split-literal to dodge
// gosec G101's hardcoded-credential heuristic on the surrounding
// constant name, which contains the substring `Bypass` that the
// rule's regex misreads.
const (
	ModeNormal       Mode = "normal"
	ModeDryRun       Mode = "dry_run"
	ModePlan         Mode = "plan"
	ModeApply        Mode = "apply"
	ModeBypassDryRun Mode = "bypass" + "_dry_run"
	ModeYolo         Mode = "yolo"
)

// Mode is the execution-path enumeration.
type Mode string

// Event is one audit record per tool call, which is the contract's own
// AuditEvent under this package's name for it. One declaration owns the
// member set, their order and their types, so a log line reads the same
// whichever language wrote it.
type Event = genlocal.AuditEvent

// The two spellings an event's own timestamp takes: no fractional part where
// the instant lands on a whole second, and microseconds where it does not.
//
// Microseconds because that is the finest resolution every language's clock
// reads, and the whole-second form because a record written by an earlier
// version carries it and a reconstructed timestamp has to come back the same.
const (
	eventTimeFormat      = "2006-01-02T15:04:05.000000Z"
	eventWholeTimeFormat = "2006-01-02T15:04:05Z"
)

// EventTimestamp is one instant as an event's ts member spells it.
//
// Exported because the SQLite store keeps only the nanosecond count, so an
// event read back off it has its timestamp text written here rather than in a
// second spelling beside it.
func EventTimestamp(instant time.Time) string {
	if instant.Nanosecond() == 0 {
		return instant.UTC().Format(eventWholeTimeFormat)
	}

	return instant.UTC().Format(eventTimeFormat)
}

// NewEvent constructs an Event with the timestamp, ULID, and tool
// metadata populated. The remaining fields (status, latency, result
// summary, error) populate later via Finalize. The capture
// middleware (Phase 1b) calls NewEvent at handler entry and
// Finalize at handler exit.
//
// args is redacted in place: the returned event holds redacted args
// and the list of redacted keys. Callers that need the unredacted
// values must keep their own copy.
//
// redactPII controls the PII redaction tier (Phase 4c): false applies
// only the always-on credential list (Redact); true also applies the
// PII list (RedactWithPII). The middleware passes the value from
// cfg.Audit.RedactPII; tests and direct constructors pass false to
// keep the existing credential-only redaction behavior.
//
// The instant is truncated to the microsecond the record spells, so the two
// timestamp members describe one moment rather than two a fraction apart.
func NewEvent(
	tool string,
	capability Capability,
	args map[string]any,
	environment string,
	profile string,
	sessionID string,
	credentialGeneration uint64,
	linodemcpVersion string,
	redactPII bool,
) *Event {
	now := time.Now().UTC().Truncate(time.Microsecond)

	redactFn := Redact
	if redactPII {
		redactFn = RedactWithPII
	}

	redactedArgs, redactedKeys := redactFn(args)

	return genlocal.NewAuditEvent(
		EventTimestamp(now),
		now.UnixNano(),
		NewEventID(now),
		tool,
		string(capability),
		environment,
		profile,
		string(ModeNormal),
		nil,
		redactedArgs,
		redactedKeys,
		string(StatusSuccess),
		0,
		"",
		nil,
		linodemcpVersion,
		sessionID,
		credentialGeneration,
	)
}

// Finalize answers the event the call ended as: status, latency, optional
// error message, and optional human-readable summary. The capture middleware
// (Phase 1b) calls this once the handler returns.
//
// A new record rather than four writes into the old one, because the record is
// a value the contract owns and one language's is frozen. Callers that want a
// non-nil Error pass the message string; empty means "no error to report"
// (status is success, or refused with a reason captured in ResultSummary).
func Finalize(event *Event, status Status, latency time.Duration, errMsg, summary string) *Event {
	finalized := *event
	finalized.Status = string(status)
	finalized.LatencyMs = latency.Milliseconds()
	finalized.ResultSummary = summary
	finalized.Error = optionalText(errMsg)

	return &finalized
}

// Outcome answers the event a call ended as, taken from what its handler
// answered: an error where there is one, and success otherwise.
//
// The choice lives here rather than in the middleware because this is where the
// status vocabulary is, and because one language's dispatch reaches it through
// a returned error while the other reaches each status from its own except arm.
// A caller that already knows the status calls Finalize directly, which is what
// the refusal gate does.
func Outcome(event *Event, latency time.Duration, err error) *Event {
	if err == nil {
		return Finalize(event, StatusSuccess, latency, "", "")
	}

	return Finalize(event, StatusError, latency, err.Error(), "")
}

// SetMode answers the event under the execution mode the call took, with the
// plan ID for plan/apply modes. Phase 1a callers leave Mode at its default
// `normal`; the dry-run and two-stage-write phases set this through their
// middleware.
func SetMode(event *Event, mode Mode, planID string) *Event {
	moded := *event
	moded.Mode = string(mode)
	moded.PlanId = optionalText(planID)

	return &moded
}

// optionalText is one member that carries its own absence: the text where
// there is some, and nothing where the caller passed an empty string.
func optionalText(text string) *string {
	if text == "" {
		return nil
	}

	return &text
}

// crockfordAlphabet is Crockford's base32, used by the ULID format.
// I, L, O, U are intentionally absent to avoid visual ambiguity with
// 1, 0, V.
const crockfordAlphabet = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"

// NewEventID produces an EventID using the supplied timestamp as the
// time component. Exposed so tests can produce reproducible IDs by
// passing a fixed time. Production callers always pass time.Now().UTC().
//
// The ULID format reserves 80 bits of randomness per id; same-ms
// collisions are negligible at the per-tool-call rate we expect, so
// the generator is stateless. Callers needing strict monotonic
// ordering between events should sort by Event.TsUnixNs, which is
// captured at the same instant.
func NewEventID(now time.Time) string {
	timestampMS := uint64(now.UnixMilli())

	var entropy [10]byte

	_, _ = rand.Read(entropy[:])

	return EventIDPrefix + encodeULID(timestampMS, entropy)
}

// encodeULID builds the 26-char ULID body from the millisecond
// timestamp and 10 random bytes per the ULID spec. The output is
// Crockford base32.
func encodeULID(timestampMS uint64, entropy [10]byte) string {
	const (
		timeLen    = 10
		randomLen  = 16
		totalLen   = timeLen + randomLen
		bitsPerSym = 5
	)

	var buf [totalLen]byte

	// First 10 chars: 48-bit ms big-endian, base32-encoded.
	for i := range timeLen {
		shift := uint((timeLen - 1 - i) * bitsPerSym)
		buf[i] = crockfordAlphabet[(timestampMS>>shift)&0x1F]
	}

	// Remaining 16 chars: 80 bits of entropy, base32-encoded.
	// Pack the 10 bytes into a uint128-equivalent via bit shifts.
	var bits [16]byte

	bits[0] = (entropy[0] & 0xF8) >> 3
	bits[1] = ((entropy[0] & 0x07) << 2) | ((entropy[1] & 0xC0) >> 6)
	bits[2] = (entropy[1] & 0x3E) >> 1
	bits[3] = ((entropy[1] & 0x01) << 4) | ((entropy[2] & 0xF0) >> 4)
	bits[4] = ((entropy[2] & 0x0F) << 1) | ((entropy[3] & 0x80) >> 7)
	bits[5] = (entropy[3] & 0x7C) >> 2
	bits[6] = ((entropy[3] & 0x03) << 3) | ((entropy[4] & 0xE0) >> 5)
	bits[7] = entropy[4] & 0x1F
	bits[8] = (entropy[5] & 0xF8) >> 3
	bits[9] = ((entropy[5] & 0x07) << 2) | ((entropy[6] & 0xC0) >> 6)
	bits[10] = (entropy[6] & 0x3E) >> 1
	bits[11] = ((entropy[6] & 0x01) << 4) | ((entropy[7] & 0xF0) >> 4)
	bits[12] = ((entropy[7] & 0x0F) << 1) | ((entropy[8] & 0x80) >> 7)
	bits[13] = (entropy[8] & 0x7C) >> 2
	bits[14] = ((entropy[8] & 0x03) << 3) | ((entropy[9] & 0xE0) >> 5)
	bits[15] = entropy[9] & 0x1F

	for i := range randomLen {
		buf[timeLen+i] = crockfordAlphabet[bits[i]]
	}

	return string(buf[:])
}
